### DISCUSSION TITLE
What are possible reasons for an include to not work in a template file?

### DISCUSSION THREAD

**Discussion by Andre601** [fralau/mkdocs-macros-plugin#255] [Q&A]
Full context: https://github.com/mkdocs/mkdocs/discussions/3888

My current issue is, that I'm unable to include a separate template file through the `include` option of Jinja2.

I've set the macros plugin's `include_dir` to `docs/.overrides/` (Can't have it outside the docs folder due to the repository's structure) and have files `api_docs/layout.md` and `api_docs/parts/constructor_summaries.md` (relative to the mentioned include directory) set.

Now, using `{% include 'api_docs/layout.md' %}` works fine when injected into the markdown through a `on_pre_page_macros` python file, but `{% include 'api_docs/parts/constructor_summaries.md %}` in the `layout.md` file itself returns a `TemplateNotFound` error.

Manually adding the include in the page is adding the content as it should, so the issue is with the template not resolving the include for some reason.

What could be possible issues?

**Comment by fralau** [comment]
First of all, I want to make sure that everyone is on the same page, about the distinction between the MkDocs's standard Jinja2 templating  engine, and the MkDocs-Macros templating engine? The first operates on HTML files, while the second operates (much earlier on the production chain), on the Markdown source, before rendering into HTML. They are completely distinct.

By default MkDocs-Macros considers the `doc_dir` as the root, because it is what makes more sense (see [documentation page](https://mkdocs-macros-plugin.readthedocs.io/en/latest/advanced/#including-external-files-in-pages)); but it can be changed, as you did, with `include_dir`. 

I have never tried to use MkDocs-Macros to do templating, though I guess that could kind of work (because HTML code is ignored by the Markdown parser).

There is an avenue I would suggest, and that is to ["trickle" values from the earlier to the later Jinja2 engine](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#trickling-values-into-the-mkdocs-theme). In that way, you could import your dynamic values in the MkDocs Jinja2 template, while using the facilities of MkDocs-Macros. Is that more less what @kamilkrzyskow had in mind?

**Reply by Andre601** [reply]
I think you missunderstand the situation?
I'm not sure what trickling values to the later jinja2 templates has anything to do with the issue I described here, which is that `include` calls return a `TemplateNotFound` error when used inside another template that gets included by the macros plugin.

**Reply by fralau** [reply]
Ah, I got it! Do you mean that in some way, a second template called within a first template does not behave exactly in the same way as the first template? In other words, its environment is not initialized in the same way?

I am not sure if that could lead you to the answer, but:

1. It may well be that the included template considers the directory of its own source file as the reference directory for relative paths, rather than the directory of the calling source file; so if that included file is in another directory, then the relative path will not work.

2. There are Jinja2 oddities that one needs to be aware of: particularly the difference between `include with context` (default) and `include without context`, as well as the difference between `include` and `import`. https://jinja.palletsprojects.com/en/stable/templates/#include.

So, when calling the template, I would not use `{% include 'api_docs/parts/constructor_summaries.md %}`, I would either try to adjust the paths accordingly, or (to avoid making computations), I would "anchor" the path to some variable (typically `include_dir`, that I know I can trust).

**Reply by Andre601** [reply]
I tried to include relative to the template's location but that wasn't successful either. Haven't checked if I have to provide the full path relative to the root of the mkdocs file to see if it works.

**Reply by fralau** [reply]
I would try to play around with `include_dir`?

```Markdown
The include file is: `{{ config.plugins.macros.include_dir }}/api_docs/parts/constructor_summaries.md`
```

**Reply by Andre601** [reply]
```
{{ no such element: mkdocs_macros.plugin.MacrosPlugin object['include_dir'] }}
```

**Reply by Andre601** [reply]
And just in case: I tested it in the page directly and as part of the template. Both have the same result...

**Reply by Andre601** [reply]
It's actually `config.plugins.macros.config.include_dir`

**Reply by Andre601** [reply]
Okay... Printing the include_dir value, it's the exact same for both template file and page (`docs/.overrides`) so the page must have the proper directory for the include.

**Reply by fralau** [reply]
@Andre601 Happy New Year! 

Did you find a satisfactory solution/workaround? (If yes would you like to share it?)

**Reply by kamilkrzyskow** [reply]
Happy New Year 🥳 

> There is an avenue I would suggest, and that is to ["trickle" values from the earlier to the later Jinja2 engine](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#trickling-values-into-the-mkdocs-theme). In that way, you could import your dynamic values in the MkDocs Jinja2 template, while using the facilities of MkDocs-Macros. Is that more less what @kamilkrzyskow had in mind?

More or less, yes. The hook would inject the "layout" into the files that need to have a static templated structure, with includes and whatnot. I always planned for either the hook or an MkDocs page to be the "orchestrator" of the templating.

I missed the part where Andre decided to use the `layout.md` as the root of the templating similar to the `base.html` in MkDocs templates. As the filename was mentioned once in the discussion together with a [completely unrelated issue](https://github.com/mkdocs/mkdocs/discussions/3888#discussioncomment-11679963), which was resolved.

[macros-with-layout.zip](https://github.com/user-attachments/files/18285764/macros-with-layout.zip)

EDIT: Turned out that I simply didn't put any data in the `page.meta`, for that specific instance, everything works as expected on my end 🤦 

~~The `include` inside the `layout.md` works, however the variables are not preserved in the context for the next included file.
The `page` object is not accessible inside the `constructor_summaries.md` if it's included inside the `layout.md`, and the `layout.md` is also included in the MkDocs page.~~

~~I haven't looked into the code deeper, since @fralau could perhaps have a quick fix for preserving context variables throughout the include chain.~~

**Reply by fralau** [reply]
@kamilkrzyskow  Just an idea: what about a global dictionary/object to which your included markdown pages could add information? That should take care of the persistence?

An easy place where to store that dictionary (possibly with some initial values) would be the `extra` part of the config file.

**Reply by kamilkrzyskow** [reply]
> Just an idea: what about a global dictionary/object to which your included markdown pages could add information? That should take care of the persistence?

Andre would have to comment on that if he's able to work with this kind of setup. However, he asked about access to `page.meta`, so I think that preserving the unique context for each page is in a way, a requirement.

Also the default `include` does pass the `context`, so there is even an expectation that variables should be preserved and even modifiable from the `leaf` level, just like it's the case of MkDocs Jinja2 templating 🤔

**Reply by fralau** [reply]
@kamilkrzyskow That is the case: the `extra` variable in MkDocs Macros is there for the purpose of storing.

Yes, the include does pass the context, and I do not reasons why it should not work; however, the the context will not survive past the _page's_ context.

Also, the `meta` variable is the way to "trickle" values to the the HTML template, and it is page by page (that is an MkDoc's design choice).

However it should be possible to work around this by storing your meta more or less:

```python
env.variables['store'] = {}

def on_post_page_macros(env):
    """
    Actions to be done after macro interpretation,
    when the markdown is already generated
    """

   # update the store variable with the meta of the page, last value overwrites the first
    env.variables['store'].update(page.meta)
```

**Reply by kamilkrzyskow** [reply]
Turned out that I simply didn't add anything to the `page.meta` for the new file. I somehow had the impression that my `hook.py` would add data to it, but it wasn't the case. So in general the chained include in a root `layout.md` does work.

So without actual files where Andre is getting his error I can't see how to find the issue. I feel like we're running in circles and it will end with some simple typo or something even simpler.

[macros-with-layout-and-copy.zip](https://github.com/user-attachments/files/18286050/macros-with-layout-and-copy.zip)

**Reply by fralau** [reply]
You might be right, for what my opinion is worth... Since MkDocs is designed for documentation, you could try to print out variables and context left and right, until you notice something unusual?

**Reply by Andre601** [reply]
Here's a zip of the docs folder, mkdocs.yml and requirements.txt I use on production with the current mkdocs-macros setup.
Currently only one page uses it which is `api/reference/api/ch.andre601.advancedserverlist.api/placeholderprovider.md`

[docs.zip](https://github.com/user-attachments/files/18292485/docs.zip)

The basic setup you gave does not raise the error, which most likely means something in my setup is messing with things... Maybe a conflicting plugin perhaps?

**Reply by kamilkrzyskow** [reply]
The `docs/.overrides` directory is missing, so I got the custom_dir from https://codeberg.org/Andre601/AdvancedServerList
I added placeholder macros templates files, which are added in the hook.
I also added a `textwrap.dedent` to the footer text.
Everything works as expected otherwise.

[macros-andre.zip](https://github.com/user-attachments/files/18293338/macros-andre.zip)

**Reply by Andre601** [reply]
> The `docs/.overrides` directory is missing,

That is weird... I simply selected the docs folder alongside the mkdocs.yml file and requirements.txt and did right-click -> "send to... -> Zip-compressed folder" on Windows.
So I guess it skipped that folder - alongside others, as you can see in the screenshot - when making the zip...

![grafik](https://github.com/user-attachments/assets/545c4881-e352-4d0d-9db8-df2d49ed4d07)

**Reply by Andre601** [reply]
Also, forgot to mention, but this version uses a different aproach where I inject the `include` statements of all template files in the page itself instead of using a main template file to include everything else from...

**Reply by kamilkrzyskow** [reply]
So was this resolved @Andre601? I have this weird feeling it didn't while it should given all the examples till now 🤔 

> That is weird... I simply selected the docs folder alongside the mkdocs.yml file and requirements.txt and did right-click -> "send to... -> Zip-compressed folder" on Windows.

I tried it myself, and only the first directory with the `.` prefix is included, if you remove `.hooks` then `.overrides` will be included. Neat edge case in Windows compressor 😆 

Therefore, it's better to rely on 3rd party toold like 7-zip that won't have the same issues, but I also often recommended the `Send to` method.

**Reply by Andre601** [reply]
Yeah, pretty sure the issue is still there.

Like I said am I using a workaround by having the individual `{% include '...' %}` be added on the page through the `on_pre_page_macros` event. It's not too easy to have but A) it works and B) the page context is available for the template files.

**Comment by Andre601** [comment]
Wanted to share what I managed to achieve with the plugin:
https://preview.andre601.ch/@AdvancedServerList~398/api/reference/

All the pages under this link (with exception to the ones listing sub-pages) are having content generated from Page meta using templates.

Templates are included using this code snippet added as a module:
```py
def on_pre_page_macros(env):
    footer = """
    
    {% if page.meta and page.meta.api %}
    {% include 'api_docs/constructor_summaries.md' %}
    {% include 'api_docs/nested_class_summaries.md' %}
    {% include 'api_docs/enum_summaries.md' %}
    {% include 'api_docs/method_summaries.md' %}
    
    {% include 'api_docs/constructor_details.md' %}
    {% include 'api_docs/enum_details.md' %}
    {% include 'api_docs/method_details.md' %}
    {% endif %}
    """
    env.markdown += footer
```

Changes are not yet merged, as I'm open for feedback on improving this (One plan is to maybe have a JSON file in the future to create the necessary stuff instead of page meta, but that's not yet clear).
You can find the PR and changes here: https://codeberg.org/Andre601/AdvancedServerList/pulls/398
